<a href="https://colab.research.google.com/github/GowreeshRaja/machine-learning-tasks/blob/main/Task_4_Decision_Tree.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task 4: Decision Tree Classifier - Diabetes Prediction

In [ ]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import (
    accuracy_score,
    confusion_matrix,
    precision_score,
    recall_score,
    f1_score
)

In [ ]:
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"

df = pd.read_csv(url)

df.head()

,6,148,72,35,0,33.6,0.627,50,1
0,1,85,66,29,0,26.6,0.351,31,0
1,8,183,64,0,0,23.3,0.672,32,1
2,1,89,66,23,94,28.1,0.167,21,0
3,0,137,40,35,168,43.1,2.288,33,1
4,5,116,74,0,0,25.6,0.201,30,0


In [ ]:
df.columns = [
    "Pregnancies",
    "Glucose",
    "BloodPressure",
    "SkinThickness",
    "Insulin",
    "BMI",
    "DiabetesPedigreeFunction",
    "Age",
    "Outcome"
]

df.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,1,85,66,29,0,26.6,0.351,31,0
1,8,183,64,0,0,23.3,0.672,32,1
2,1,89,66,23,94,28.1,0.167,21,0
3,0,137,40,35,168,43.1,2.288,33,1
4,5,116,74,0,0,25.6,0.201,30,0


In [ ]:
df.isnull().sum()

,0
Pregnancies,0
Glucose,0
BloodPressure,0
SkinThickness,0
Insulin,0
BMI,0
DiabetesPedigreeFunction,0
Age,0
Outcome,0


In [ ]:
zero_counts = (df == 0).sum()

zero_counts

,0
Pregnancies,111
Glucose,5
BloodPressure,35
SkinThickness,227
Insulin,373
BMI,11
DiabetesPedigreeFunction,0
Age,0
Outcome,500


In [ ]:
columns_to_replace = [
    "Glucose",
    "BloodPressure",
    "SkinThickness",
    "Insulin",
    "BMI"
]

df[columns_to_replace] = df[columns_to_replace].replace(0, np.nan)

df.isnull().sum()

,0
Pregnancies,0
Glucose,5
BloodPressure,35
SkinThickness,227
Insulin,373
BMI,11
DiabetesPedigreeFunction,0
Age,0
Outcome,0


In [ ]:
for column in columns_to_replace:
    df[column] = df[column].fillna(df[column].median())

df.isnull().sum()

,0
Pregnancies,0
Glucose,0
BloodPressure,0
SkinThickness,0
Insulin,0
BMI,0
DiabetesPedigreeFunction,0
Age,0
Outcome,0


In [ ]:
X = df.drop("Outcome", axis=1)
y = df["Outcome"]

print("Features shape:", X.shape)
print("Target shape:", y.shape)

Features shape: (767, 8)
Target shape: (767,)


In [ ]:
print("Number of rows:", len(df))
print("Number of columns:", len(df.columns))

df.shape

Number of rows: 767
Number of columns: 9


(767, 9)

In [ ]:
original_df = pd.read_csv(url)

print("Original dataset shape:", original_df.shape)
print("Current dataset shape:", df.shape)

Original dataset shape: (767, 9)
Current dataset shape: (767, 9)


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print("Training set:", X_train.shape)
print("Testing set:", X_test.shape)

Training set: (613, 8)
Testing set: (154, 8)


In [ ]:
dt_model = DecisionTreeClassifier(random_state=42)

dt_model.fit(X_train, y_train)

y_pred = dt_model.predict(X_test)

print("Decision Tree model trained successfully.")

Decision Tree model trained successfully.


In [ ]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)

print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

print("\nConfusion Matrix:")
print(cm)

Accuracy: 0.7077922077922078
Precision: 0.6111111111111112
Recall: 0.5789473684210527
F1-score: 0.5945945945945946

Confusion Matrix:
[[76 21]
 [24 33]]


In [ ]:
dt_model_depth3 = DecisionTreeClassifier(
    max_depth=3,
    random_state=42
)

dt_model_depth3.fit(X_train, y_train)

y_pred_depth3 = dt_model_depth3.predict(X_test)

print("Restricted Decision Tree trained successfully.")

Restricted Decision Tree trained successfully.


In [ ]:
accuracy_depth3 = accuracy_score(y_test, y_pred_depth3)
precision_depth3 = precision_score(y_test, y_pred_depth3)
recall_depth3 = recall_score(y_test, y_pred_depth3)
f1_depth3 = f1_score(y_test, y_pred_depth3)
cm_depth3 = confusion_matrix(y_test, y_pred_depth3)

print("Accuracy:", accuracy_depth3)
print("Precision:", precision_depth3)
print("Recall:", recall_depth3)
print("F1-score:", f1_depth3)

print("\nConfusion Matrix:")
print(cm_depth3)

Accuracy: 0.7467532467532467
Precision: 0.65
Recall: 0.6842105263157895
F1-score: 0.6666666666666666

Confusion Matrix:
[[76 21]
 [18 39]]


In [ ]:
comparison = pd.DataFrame({
    "Metric": ["Accuracy", "Precision", "Recall", "F1-score"],
    "Decision Tree": [accuracy, precision, recall, f1],
    "Decision Tree (max_depth=3)": [
        accuracy_depth3,
        precision_depth3,
        recall_depth3,
        f1_depth3
    ]
})

comparison

,Metric,Decision Tree,Decision Tree (max_depth=3)
0,Accuracy,0.707792,0.746753
1,Precision,0.611111,0.650000
2,Recall,0.578947,0.684211
3,F1-score,0.594595,0.666667


In [ ]:
feature_importance = pd.DataFrame({
    "Feature": X.columns,
    "Importance": dt_model_depth3.feature_importances_
})

feature_importance = feature_importance.sort_values(
    by="Importance",
    ascending=False
)

feature_importance

,Feature,Importance
1,Glucose,0.570905
5,BMI,0.196821
7,Age,0.150482
4,Insulin,0.081792
3,SkinThickness,0.000000
2,BloodPressure,0.000000
0,Pregnancies,0.000000
6,DiabetesPedigreeFunction,0.000000


## Feature Importance Interpretation

The feature importance results show that **Glucose** has the highest importance in the Decision Tree model, with an importance value of approximately 0.571.

**BMI** is the second most important feature, followed by **Age** and **Insulin**.

The remaining features — **SkinThickness, BloodPressure, Pregnancies, and DiabetesPedigreeFunction** — have an importance value of 0 in this particular Decision Tree.

This indicates that the model relied mainly on Glucose, BMI, Age, and Insulin when making its predictions.

## Conclusion

Two Decision Tree models were trained and evaluated for diabetes prediction.

The first Decision Tree achieved an accuracy of approximately **70.78%**, while the Decision Tree with `max_depth=3` achieved an accuracy of approximately **74.68%**.

The restricted-depth model also achieved higher precision, recall, and F1-score on the test set.

The feature importance analysis showed that **Glucose** was the most important feature, followed by **BMI, Age, and Insulin**.

Therefore, both models were successfully evaluated and compared, and the important features used by the restricted Decision Tree were identified.